File: typescript/src/services/query-builder.ts
import { DatasetQueryResponseSchema, DatasetDocument } from '../models/dataset';
import type { DatasetService } from './datasets';

export class QueryBuilder {
  private _service: DatasetService;
  private _dataset: string;
  private _searchTerm?: string;
  private _filters: Record<string, any> = {};
  private _fields?: string[];
  private _limit: number = 100;
  private _offset?: number;
  private _cursor?: string;

  constructor(service: DatasetService, dataset: string) {
    this._service = service;
    this._dataset = dataset;
  }

  search(term: string): this {
    this._searchTerm = term;
    return this;
  }

  where(key: string, value: any): this {
    this._filters[key] = value;
    return this;
  }

  fields(projection: string[]): this {
    this._fields = projection;
    return this;
  }

  limit(count: number): this {
    this._limit = count;
    return this;
  }

  offset(n: number): this {
    this._offset = n;
    return this;
  }

  after(cursor: string): this {
    this._cursor = cursor;
    return this;
  }

  async execute(): Promise<ReturnType<typeof DatasetQueryResponseSchema.parse>> {
    return this._service.query(this._dataset, {
      search_term: this._searchTerm,
      filters: this._filters,
      fields: this._fields,
      limit: this._limit,
      offset: this._offset,
      cursor: this._cursor,
    });
  }

  /**
   * Returns an async generator for streaming results.
   */
  async *stream(): AsyncIterableIterator<DatasetDocument> {
    let currentCursor = this._cursor;

    while (true) {
      const result = await this._service.query(this._dataset, {
        search_term: this._searchTerm,
        filters: this._filters,
        fields: this._fields,
        limit: this._limit,
        offset: this._offset,
        cursor: currentCursor,
      });

      for (const doc of result.data) {
        yield doc;
      }

      if (!result.next_cursor) break;
      currentCursor = result.next_cursor;
    }
  }
}

File: go/query_builder.go
package dataflare

import (
	"context"

	"github.com/dataflarelab/df-sdk/go/models"
)

// QueryBuilder provides a fluent interface for building dataset queries.
type QueryBuilder struct {
	service *DatasetService
	dataset string
	search  string
	filters map[string]interface{}
	fields  []string
	limit   int
	offset  int
}

// NewQueryBuilder creates a new QueryBuilder instance.
func NewQueryBuilder(service *DatasetService, dataset string) *QueryBuilder {
	return &QueryBuilder{
		service: service,
		dataset: dataset,
		filters: make(map[string]interface{}),
	}
}

// Search sets the search term for the query.
func (b *QueryBuilder) Search(term string) *QueryBuilder {
	b.search = term
	return b
}

// Where adds a filter to the query.
func (b *QueryBuilder) Where(key string, value interface{}) *QueryBuilder {
	b.filters[key] = value
	return b
}

// Fields sets the fields to include in the query results.
func (b *QueryBuilder) Fields(fields []string) *QueryBuilder {
	b.fields = fields
	return b
}

// Limit sets the maximum number of results to return.
func (b *QueryBuilder) Limit(limit int) *QueryBuilder {
	b.limit = limit
	return b
}

// Offset sets the number of results to skip.
func (b *QueryBuilder) Offset(offset int) *QueryBuilder {
	b.offset = offset
	return b
}

// Params returns the query parameters as QueryOptions.
func (b *QueryBuilder) Params() *models.QueryOptions {
	return &models.QueryOptions{
		SearchTerm: b.search,
		Filters:    b.filters,
		Fields:     b.fields,
		Limit:      b.limit,
		Offset:     b.offset,
	}
}

// Execute executes the built query.
func (b *QueryBuilder) Execute(ctx context.Context) (*models.DatasetResponse, error) {
	return b.service.Query(ctx, b.dataset, b.Params())
}

// Stream performs a streaming paginated query on a dataset.
func (b *QueryBuilder) Stream(ctx context.Context) (<-chan models.Document, <-chan error) {
	return b.service.Stream(ctx, b.dataset, b.Params())
}
